### ISSUE TITLE
模型蒸馏实践

### ISSUE THREAD

**Issue by xcodebuild** [xcodebuild/blog#1]
## 背景
我们经常听到蒸馏（distillation）这个词的场景，通常是某个大模型厂商指责另外一个模型厂商通过蒸馏他们的模型来提高模型能力。

事实上蒸馏在我们实际的适用场景中是一个非常实用的技术，它可以让我们低成本的让小参数量模型（例如 `Qwen2.5-0.5B`）具备大参数量通用模型部分特定的能力，本文用一个具体的场景来实践这种技术，并介绍在真实场景中它是如何帮我们降低成本、提高性能的。

## 典型场景

我们经常使用大模型来处理一些垂直需求，例如日志打标、关键信息抽取、场景识别之类的。

我们用一个在电商 App 中常用的功能：`从文本提取接结构化的地址信息`举例
```
输入：
李明 13800138000 上海市浦东新区世纪大道100号上海环球金融中心72楼
输出：
{
  "name": "李明",
  "phone": "13800138000",
  "province": "上海市",
  "city": "上海市",
  "district": "浦东新区",
  "specific_location": "世纪大道100号上海环球金融中心72楼"
}
```

我们可以用简单的代码直接把上面的 `few-shot` 给到 `Qwen/Qwen2.5-0.5B-Instruct`。

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model_name = "Qwen/Qwen2.5-0.5B-Instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    torch_dtype=torch.float16,
    device_map="auto"
)

def build_messages(address):
    messages = [
        {"role": "system", "content": """
你是一个地址解析专家。请将用户提供的中文地址解析为结构化JSON，只输出JSON本身，不要有其他解释。例子
输入：李明 13800138000 上海市浦东新区世纪大道100号上海环球金融中心72楼
输出：
{
  "name": "李明",
  "phone": "13800138000",
  "province": "上海市",
  "city": "上海市",
  "district": "浦东新区",
  "specific_location": "世纪大道100号上海环球金融中心72楼"
}
"""},
        {"role": "user", "content": f"地址：{address}"}
    ]
    text = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    return text

test_address = "我的地址是 AKA 浙江省杭州市余杭区文一西路969号阿里巴巴西溪园区B区6号楼 13840393849"
prompt = build_messages(test_address)
inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

outputs = model.generate(
    **inputs,
    max_new_tokens=128,
    do_sample=False,          # 关闭采样，避免重复
    temperature=None,         # 不使用temperature
    top_p=None,
    pad_token_id=tokenizer.eos_token_id
)

response = tokenizer.decode(outputs[0][inputs.input_ids.shape[1]:], skip_special_tokens=True)
print(response)
```

得到的结果是

```
{
  "name": "阿里巴巴西溪园区B区6号楼",
  "phone": "13840393849",
  "province": "浙江省",
  "city": "杭州市",
  "district": "余杭区",
  "specific_location": "文一西路969号"
}
```

我们会发现对于简单的例子，小尺寸模型能够解析，但是稍微复杂一些的例子（顺序变化、混入了其他无关描述）就无法正确解析了。
接下来我们来把教师模型（这里采用 `deepseek-v3.2`）的地址信息提取能力蒸馏到 `qwen2.5-0.5B` 上。

## 蒸馏数据

因为这里的数据构造比较简单，就使用教师模型生成一系列的输入输出数据，用于后续的模型微调。

对于真实的适用场景，我们可能需要更复杂的方式来构造数据，但核心思想是不变的，即用教师模型来生产高质量的数据用于微调。

```
你是一个数据生成器。请生成100条中文物流信息样本，每条样本包含姓名、电话、地址（省市区+详细地址）。要求：
1. 姓名：中文名、英文名、网络昵称、简单字母数字拼接等都有可能
2. 电话：11位手机号
3. 地址：真实风格，包括省市区街道门牌号及POI
4. 输出格式为JSONL，每行包含两个字段：input 和 output
   - input：一段自然语言文本，可以是“姓名+电话+地址”的不同组合顺序，也可以带有一些无关的对话信息等干扰
   - output：包含 name, phone, province, city, district, specific_location 的JSON。

示例：
{"input": "李明 13800138000 上海市浦东新区世纪大道100号上海环球金融中心72楼", 
 "output": {"name": "李明", "phone": "13800138000", "province": "上海市", "city": "上海市", "district": "浦东新区", "specific_location": "世纪大道100号上海环球金融中心72楼"}}
{"input": "收货人：王芳，电话13912345678，地址浙江省杭州市西湖区学院路78号阿里中心", 
 "output": {"name": "王芳", "phone": "13912345678", "province": "浙江省", "city": "杭州市", "district": "西湖区", "specific_location": "学院路78号阿里中心"}}
{"input": "我的地址是AKA浙江省杭州市余杭区文一西路969号阿里巴巴西溪园区B区6号楼 13840393849", 
 "output": {"name": "", "phone": "13840393849", "province": "浙江省", "city": "杭州市", "district": "余杭区", "specific_location": "文一西路969号阿里巴巴西溪园区B区6号楼"}}
```

这样会生成 100 条数据，存储到 `data.jsonl`。

## 微调

我们用生成的数据对模型进行微调，这里偷个懒不展开写训练的代码了，直接让 `opencode` 写：

```
用 data.jsonl 的数据对 Qwen/Qwen2.5-0.5B-Instruct 模型进行微调
```

微调产生新的模型后，重新测试：

```
输入：
我的地址是 AKA 浙江省杭州市余杭区文一西路969号阿里巴巴西溪园区B区6号楼 13840393849
输出：
{"name":"KAHAN","phone":"13840393849","province":"浙江省","city":"杭州市","district":"余杭区","specific_location":"文一西路969号阿里巴巴西溪园区B区6号楼"}
```

可以看到除了 `name` 还有一些小问题外，其他的表现已经很好了（这里其实是因为 `AKA` 这样的名字在训练数据中很少见）。

其他的例子也表现得更稳定，例如：

```
=== 输入地址 ===
收件人：王力宏，电话13900001111，地址北京市朝阳区望京街10号SOHO T3 2601

=== 原模型输出 ===
{
  "name": "王力宏",
  "phone": "13900001111",
  "address": "北京市朝阳区望京街10号SOHO T3 2601"
}

=== 微调后模型输出 ===
{"name":"王力宏","phone":"13900001111","province":"北京市","city":"北京市","district":"朝阳区","specific_location":"望京街10号SOHO T3 2601"}
```

## 总结
在日常使用中，`0.5B` 的小尺寸模型本身的能力很难满足诉求，但大尺寸的模型成本又高、推理又慢。针对小尺寸模型在这些垂直场景构造一系列的高质量数据，可以通过微调提高它在特定场景的表现和稳定性。

而蒸馏就是直接借助通用大模型构造高质量数据，从而实现把通用大模型的部分能力迁移到小尺寸模型上，实现高性能、低成本、表现好的模型效果。

## 备注
> 一些过程中需要了解的其他知识
- 微调：可以简单理解为通过少量数据调整一个已经存在的模型的权重，通常我们使用的是 LoRA（例如上面的例子中），他是在原有的权重矩阵边附加低秩矩阵，不用修改原模型。整体需要的资源显著更少，训练更快。
- loss：微调的过程中我们会看到 loss 相关的输出，其衡量的是模型推理结果的错误程度（损失）
    - train_loss：模型在训练集上的损失，反应模型对训练数据的拟合程度
    - eval_loss：模型在验证集上的损失，反应模型运用（泛化）的能力
    - eval_loss 远大于 train_loss：说明模型记住了训练集，但实际使用表现不好，可能是过拟合
    - 两者都高：说明模型未能有效学到规律
    - train_loss 和 eval_loss 都不断降低后趋于稳定（收敛）：说明模型有效学习了规律
